wrangler-core/src/main/java/io/cdap/directives/column/SetType.java

```diff
@@ -121,8 +130,38 @@ public Schema getOutputSchema(SchemaResolutionContext context) {
         .map(
           field -> {
             try {
-              return field.getName().equals(col) ?
-                Schema.Field.of(col, ColumnConverter.getSchemaForType(type, scale)) : field;
+              if (field.getName().equals(col)) {
+                Schema intermediateSchema = field.getSchema().getNonNullable();
+                Pair<Integer, Integer> scaleAndPrecision = findPrecisionAndScale(intermediateSchema);
+                Integer inputSchemaScale = scaleAndPrecision.getSecond();
+                Integer inputSchemaPrecision = scaleAndPrecision.getFirst();
+
+                if (scale == null && precision == null) {
+                  return Schema.Field.of(col, ColumnConverter.getSchemaForType(type,
+                      inputSchemaScale, inputSchemaPrecision));
+                } else if (scale == null && inputSchemaScale != null) {
+                  if (precision - inputSchemaScale < 1) {
+                    throw new DirectiveParseException(String.format(
+                        "Cannot set scale as '%s' and precision as '%s' when "
+                            + "given precision - scale is less than 1 ", inputSchemaScale,
+                        precision));
+                  }
+                  return Schema.Field.of(col, ColumnConverter.getSchemaForType(type,
+                      inputSchemaScale, precision));
+                } else if (precision == null && inputSchemaPrecision != null) {
+                  if (inputSchemaPrecision - scale < 1) {
+                    throw new DirectiveParseException(String.format(
+                        "Cannot set scale as '%s' and precision as '%s' when "
+                            + "given precision - scale is less than 1 ", scale,
+                        inputSchemaPrecision));
+                  }
+                  return Schema.Field.of(col, ColumnConverter.getSchemaForType(type,
```

Comment:
It would be better to set these values in outputScale, outputPrecision variables and have just a single return statement

Reply:
Done